### PR TITLE
Support listening for listener changes

### LINKED ISSUE #46: Support listening for listener changes
EventEmitter's implementation:

```js
const EventEmitter = require("events")

const emitter = new EventEmitter()

emitter.on("newListener", (data, callback) => {

})

emitter.on("removeListener", (data, callback) => {

})
```

cc @sindresorhus

### PR DESCRIPTION
**Pull request by Richienb**
Closes #46

cc @sindresorhus

### CODE REVIEW

**Comment by sindresorhus** [comment]
The `symbol` should be `unique symbol`. And don’t do unrelated style changes.

**Comment by Richienb** [comment]
@sindresorhus Somehow, xo didn't catch that.

**Comment by stroncium** [comment]
Docs definitely need to say that it doesn't trigger on listener changes for meta events.

**Comment by Richienb** [comment]
@sindresorhus Please re-review.

File: readme.md [hunk 1]
```diff
@@ -56,6 +56,15 @@ Returns an unsubscribe method.
 
 Using the same listener multiple times for the same event will result in only one method call per emitted event.
 
+##### Custom subscribable events
+
+Emittery exports some symbols which represent custom events that can be passed to `Emitter.on` and similar functions.
+
+- `Emittery.listenerAdded` - Fires when an event listener was added. An object with `eventName` and `listener` is provided as data where possible.
+- `Emittery.listenerRemoved` - Fires when an event listener was removed. An object with `eventName` and `listener` is provided as data where possible.
```
Comment (stroncium) [inline readme.md]: "is provided as data where possible" - `listener` is always there, and `eventName` is only not there in case on `onAny`/`offAny`. Need clear docs on both cases.

Comment (Richienb) [reply readme.md]: Fixed!

Comment (stroncium) [reply readme.md]: Still no docs on what happens when `onAny`/`offAny` is used.

Comment (Richienb) [reply readme.md]: @stroncium Fixed!

File: test/index.js [hunk 2]
```diff
@@ -27,6 +27,58 @@ test('on() - symbol eventName', async t => {
 	t.deepEqual(calls, [1, 2]);
 });
 
+test.cb('on() - listenerAdded', t => {
```
Comment (sindresorhus) [inline test/index.js]: Generally, don't use `test.cb`. There's usually a better solution. Here, you could use https://github.com/sindresorhus/p-event

File: readme.md [hunk 3]
```diff
@@ -56,6 +56,20 @@ Returns an unsubscribe method.
 
 Using the same listener multiple times for the same event will result in only one method call per emitted event.
 
+##### Custom subscribable events
+
+Emittery exports some symbols which represent custom events that can be passed to `Emitter.on` and similar functions.
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
Emittery exports some symbols which represent custom events that can be passed to `Emitter.on` and similar methods.
```

File: index.d.ts [hunk 4]
```diff
@@ -18,6 +18,20 @@ declare class Emittery {
 	*/
 	static mixin(emitteryPropertyName: string, methodNames?: readonly string[]): Function;
 
+	/**
+	Fires when an event listener was added.
+	
+	An object with `listener` and `eventName` (if `on` or `off` was used) is provided.
+	*/
+	static readonly listenerAdded: unique symbol;
```
Comment (sindresorhus) [inline index.d.ts]: Can you add a usage example to this and `listenerRemoved`? Readme too.

File: index.d.ts [hunk 5]
```diff
@@ -18,6 +18,20 @@ declare class Emittery {
 	*/
 	static mixin(emitteryPropertyName: string, methodNames?: readonly string[]): Function;
 
+	/**
+	Fires when an event listener was added.
+	
+	An object with `listener` and `eventName` (if `on` or `off` was used) is provided.
```
Comment (sindresorhus) [inline index.d.ts]: "provided" where? Not clear enough.

File: test/index.js (hunk 2 above)
Comment (Richienb) [reply test/index.js]: Fixed!

File: index.d.ts (hunk 5 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts (hunk 4 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

**Comment by Richienb** [comment]
@sindresorhus Ready for re-review!

File: test/index.js [hunk 6]
```diff
@@ -25,14 +26,66 @@ test('on() - symbol eventName', async t => {
...
 	const emitter = new Emittery();
 
-	emitter.on('string', () => {});
-	emitter.on(Symbol('symbol'), () => {});
+	emitter.on('string', () => { });
+	emitter.on(Symbol('symbol'), () => { });
 
 	t.throws(() => {
-		emitter.on(42, () => {});
+		emitter.on(42, () => { });
```
Comment (sindresorhus) [inline test/index.js]: Don't do unrelated changes...

File: readme.md [hunk 7]
```diff
@@ -53,6 +53,38 @@ Returns an unsubscribe method.
...
+	// Handle data
+});
+```
+
+###### Listener data
+
+- `listener` - The listener that was added.
+- `eventName` - The name of the event that was added if `on` or `off` was used instead of `onAny` or `offAny`.
+
+Only events that are not of this type are able to trigger these events.
```
Comment (sindresorhus) [inline readme.md]: This is missing from the TS file.

**Comment by sindresorhus** [comment]
Could we make the event data strongly-typed? I'm thinking an overload on `.on(` and `.off(` that when one of the symbols are used, the event data is explicitly typed.

File: test/index.js (hunk 6 above)
Comment (Richienb) [reply test/index.js]: Fixed!

File: readme.md (hunk 7 above)
Comment (Richienb) [reply readme.md]: Fixed!

**Comment by Richienb** [comment]
@sindresorhus Done!

File: index.d.ts [hunk 8]
```diff
@@ -5,6 +5,12 @@ Symbol event names can be used to avoid name collisions when your classes are ex
 */
 type EventName = string | symbol;
 
+/** The data provided as `eventData` when listening for `Emittery.listenerAdded` or `Emittery.listenerRemoved`. */
+interface EventListenerData {
+	listener: (eventData?: unknown) => void;
+	eventName?: string;
```
Comment (sindresorhus) [inline index.d.ts]: Why is it optional? And it should be:
```suggestion
	eventName?: EventName;
```

File: index.d.ts [hunk 9]
```diff
@@ -85,6 +91,7 @@ declare class Emittery {
 	@returns An unsubscribe method.
 	*/
 	on(eventName: EventName, listener: (eventData?: unknown) => void): Emittery.UnsubscribeFn;
+	on(eventName: typeof Emittery.listenerAdded | typeof Emittery.listenerRemoved, listener: (eventData: EventListenerData) => void): Emittery.UnsubscribeFn
```
Comment (sindresorhus) [inline index.d.ts]: Can you add a type test to ensure this overload works correctly?

File: index.d.ts [hunk 10]
```diff
@@ -5,6 +5,12 @@ Symbol event names can be used to avoid name collisions when your classes are ex
 */
 type EventName = string | symbol;
 
+/** The data provided as `eventData` when listening for `Emittery.listenerAdded` or `Emittery.listenerRemoved`. */
+interface EventListenerData {
```
Comment (sindresorhus) [inline index.d.ts]: The name is too generic. And the interface should be exported.

File: index.d.ts [hunk 11]
```diff
@@ -5,6 +5,21 @@ Symbol event names can be used to avoid name collisions when your classes are ex
 */
 type EventName = string | symbol;
 
+/**
+The data provided as `eventData` when listening for `Emittery.listenerAdded` or `Emittery.listenerRemoved`.
+*/
+interface EventListenerData {
+	/**
+	The listener that was added.
```
Comment (sindresorhus) [inline index.d.ts]: "or removed"

File: index.d.ts [hunk 12]
```diff
@@ -5,6 +5,21 @@ Symbol event names can be used to avoid name collisions when your classes are ex
...
+	*/
+	listener: (eventData?: unknown) => void;
+
+	/**
+	The name of the event that was added if `on` or `off` was used instead of `onAny` or `offAny`.
```
Comment (sindresorhus) [inline index.d.ts]: "or removed"

File: index.d.ts (hunk 8 above)
Comment (Richienb) [reply index.d.ts]: @sindresorhus When using `emitter.onAny`, an eventName is not specified as part of the function.

File: index.d.ts (hunk 10 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts (hunk 12 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts (hunk 11 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts (hunk 9 above)
Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts (hunk 8 above)
Comment (sindresorhus) [reply index.d.ts]: > When using emitter.onAny, an eventName is not specified as part of the function.

That should be properly documented in the docs (TS/readme).

File: index.test-d.ts [hunk 13]
```diff
@@ -15,6 +15,8 @@ import Emittery = require('.');
 	ee.on('anEvent', () => Promise.resolve());
 	ee.on('anEvent', data => undefined);
 	ee.on('anEvent', data => Promise.resolve());
+	ee.on(Emittery.listenerAdded, ({eventName, listener}) => {})
+	ee.on(Emittery.listenerRemoved, ({eventName, listener}) => {})
```
Comment (sindresorhus) [inline index.test-d.ts]: Semicolons

File: index.d.ts (hunk 8 above)
Comment (Richienb) [reply index.d.ts]: @sindresorhus The docs literally say:
> The name of the event that was added or removed if `on` or `off` was used instead of `onAny` or `offAny`.

File: index.test-d.ts (hunk 13 above)
Comment (Richienb) [reply index.test-d.ts]: Fixed!

File: index.d.ts [hunk 14]
```diff
@@ -208,6 +264,21 @@ declare namespace Emittery {
...
+	The data provided as `eventData` when listening for `Emittery.listenerAdded` or `Emittery.listenerRemoved`.
+	*/
+	interface ListenerChangedData {
+		/**
+		The listener that was added or removed.
+		*/
+		listener: (eventData?: unknown) => void;
+
+		/**
+		The name of the event that was added or removed if `on` or `off` was used instead of `onAny` or `offAny`.
```
Comment (sindresorhus) [inline index.d.ts]: Re https://github.com/sindresorhus/emittery/pull/49#discussion_r389500443. I don't think it's clear enough.

```suggestion
		The name of the event that was added or removed if `.on()` or `.off()` was used, or `undefined` if `.onAny()` or `.offAny()` was used.
```

Comment (Richienb) [reply index.d.ts]: Fixed!

File: index.d.ts [hunk 15]
```diff
@@ -208,6 +264,21 @@ declare namespace Emittery {
...
+	interface ListenerChangedData {
+		/**
+		The listener that was added or removed.
+		*/
+		listener: (eventData?: unknown) => void;
+
+		/**
+		The name of the event that was added or removed if `.on()` or `.off()` was used, or `undefined` if `.onAny()` or `.offAny()` was used.
+		*/
+		eventName?: EventName;
```
Comment (sindresorhus) [inline index.d.ts]: I just realized that we don't add the `on(eventName: typeof Emittery.listenerAdded | typeof Emittery.listenerRemoved` overload to `onAny`/`offAny`, so they can never receive this TS interface, so `eventName` doesn't need to be optional.

Comment (Richienb) [reply index.d.ts]: @sindresorhus We are responding to `.onAny`, not using it directly.

```js
const Emittery = require(".")

const emitter = new Emittery()

emitter.on(Emittery.listenerAdded, ({listener, eventName}) => {
	console.log(listener)
	//=> [Function (anonymous)]

	console.log(eventName)
	//=> undefined
})

emitter.onAny(() => {
	// Process logic...
})
```

Comment (sindresorhus) [reply index.d.ts]: Right, never mind. I was confused.
